src/box/lua/load_cfg.lua

```diff
@@ -662,43 +699,58 @@ local function compare_cfg(cfg1, cfg2)
     return table.equals(cfg1, cfg2)
 end
 
+local function try_rollback_option(oldcfg, key)
+    local result, err = pcall(dynamic_cfg_revert[key])
+    if not result then
+        log.error("failed to revert '%s' configuration option: %s", key, err)
+        -- Setting to this special value should not fail
+        rawset(oldcfg, key, default_cfg_on_revert[key])
+        assert(pcall(dynamic_cfg_revert[key]))
+    end
+end
+
 local function reload_cfg(oldcfg, cfg)
     cfg = upgrade_cfg(cfg, translate_cfg)
     local newcfg = prepare_cfg(cfg, default_cfg, template_cfg,
                                module_cfg, modify_cfg)
     local ordered_cfg = {}
+    local modules_cfg = {}
     -- iterate over original table because prepare_cfg() may store NILs
-    for key, val in pairs(cfg) do
-        if dynamic_cfg[key] == nil and oldcfg[key] ~= val then
-            box.error(box.error.RELOAD_CFG, key);
+    for key in pairs(cfg) do
+        if not compare_cfg(oldcfg[key], newcfg[key]) then
+            local module = option2module[key]
+            if module == nil then
+                box.error(box.error.RELOAD_CFG, key);
+            end
+            if modules_cfg[module] == nil then
+                modules_cfg[module] = {}
+                table.insert(ordered_cfg, module)
+            end
+            modules_cfg[module][key] = true
         end
-        table.insert(ordered_cfg, key)
     end
     table.sort(ordered_cfg, sort_cfg_cb)
-    for _, key in pairs(ordered_cfg) do
-        local val = newcfg[key]
-        local oldval = oldcfg[key]
-        if not compare_cfg(val, oldval) then
-            rawset(oldcfg, key, val)
-            local result, err = pcall(dynamic_cfg[key])
-            if not result then
+    for _, module in pairs(ordered_cfg) do
+        local oldvals = {}
+        for key in pairs(modules_cfg[module]) do
```

Comment:
I don't like the name of this local variable - it's easy to confuse with `module_cfg`. Do we really need it though? Maybe, iterator over `module_config[module].options` instead?

Reply:
On f2f discussion decided to give better names to `modules_cfg` and `ordered_cfg`.